tools/flowc/incremental/fisubtype.flow

```diff
@@ -294,3 +306,100 @@ expandFiUnion(env : FcGlobalNames, union : FcTypeUnion, acc : Set<string>) -> Se
 		}
 	});
 }
+
+isFiSubType2(names : FiGlobalNames, small : FiType, big : FiType, flags : FiSubtypeFlags) -> bool {
+    flowok = \ -> 
+		(flags.allowFlow && small == FiTypeFlow());
+	args = \s : [FiType], b : [FiType] -> {
+		length(s) == length(b) && foldi(s, true, \i, acc, sa ->
+			acc && isFiSubType2(names, sa, b[i], flags)
+		);
+	}
+	fiEqualTypes(small, big) ||
+    switch (small) {
+		FiTypeParameter(__): true; // We just accept this
+		default: {
+			switch (big) {
+				FiTypeVoid():   false;
+				FiTypeBool():   flowok();
+				FiTypeInt():    flowok();
+				FiTypeDouble(): flowok();
+				FiTypeString(): flowok();
+				FiTypeArray(bt): switch (small) {
+					FiTypeArray(st): {
+						flags.allowArrays && isFiSubType2(names, st, bt, flags);
+					}
+					default: flowok();
+				}
+				FiTypeFlow():   flags.allowFlow;
+				FiTypeNative(): flowok();
+				FiTypeFunction(bargs, breturnType): {
+					switch (small) {
+						FiTypeFunction(sargs, sreturnType): {
+							isFiSubType2(names, sreturnType, breturnType, flags)
+							// Here, we reverse
+							&& args(map(bargs, \ba -> ba.type), map(sargs, \sa -> sa.type))
+						}
+						default: flowok();
+					}
+				}
+				FiTypeRef(bt): switch (small) {
+					FiTypeRef(st): {
+						isFiSubType2(names, st, bt, FiSubtypeFlags(flags with allowFlow = false))
+					}
+					default: false;
+				}
+				FiTypeParameter(__): true; // We just accept this
+				FiTypeName(bname, btypeparameters): {
+					switch (small) {
+						FiTypeName(sname, stypeparameters): {
+							typarsok = \stypars, btypars -> stypars == [] || btypars == [] || args(stypars, btypars);
+							if (sname == bname) {
+								// This is required for some cases, like switch, where we do not have typars
+								// as well as for implicit polymorphism due to polymorphic unions
+								flags.ignoreTypeparsForTypenames || typarsok(stypeparameters, btypeparameters);
+							} else {
+								flags.allowUnions && (!flags.checkNamed ||
+									(
+										typarsok(stypeparameters, btypeparameters) && 
+										getFiTypenameSubType(names, sname, bname)
+									)
+								);
+							}
+						}
+						default: flowok();
+					}
+				}
+			}
+		}
+	}
+}
+
+getFiTypenameSubType(names : FiGlobalNames, sname : string, bname : string) -> bool {
+	(sname == bname) ||
+	switch (lookupTree(names.unions, sname)) {
+		Some(small_union): {
+			switch (lookupTree(names.unions, bname)) {
+				Some(big_union): {
+					forall(small_union.typenames, \tn1 -> 
```

Comment:
i see a lot of forall, exists, etc. are you sure it works fast?

Reply:
If it is a bottleneck - I'll be bothered with it, otherwise it's more important if it works right (i.e. has a proper semantics).